### ISSUE TITLE
need to allow anonymous bind to browse an ldap repository

### ISSUE THREAD

**Issue by donraikes** [ldapjs/node-ldapjs#703]
I am porting an old php application to nodejs, and in the old application it performs an anonymous bind to our ldap server to be able to retrieve cn , sn, email, and also be able to locate upstream managers.
Based on the documentation I have seen for ldapjs, there is no way to perform an anonymous bind.

Is this possible, and if so how is it done, If not can it be added?

**Comment by UziTech** [comment]
According to the [docs](http://ldapjs.org/guide.html#bind)

> You can bind any number of times on a single connection and change that identity. Technically, it's optional, and you can support anonymous operations from clients

> (by default ldapjs will have a DN of `cn=anonymous` if the client didn't bind).

**Comment by donraikes** [comment]
So if i don't explicitly perform a bind, then the client is bound anonymously?

I have written a little test script to see if I get anything back from ldap:

```
console.log("Starting script.");
var assert = require('assert');
var ldap = require('ldapjs');
var client = ldap.createClient({
  url: 'ldap://ldap.example.com',
  reconnect: true
});
console.log("after createClient");

var opts = {
  filter: '(email=john.doe@example.com)',
  scope: 'sub',
};
client.search('dc=example,dc=com', opts, function(err, res) {
  assert.ifError(err);

    res.on('searchEntry', function(entry) {
	    console.log('entry: ' + JSON.stringify(entry.object));
		  });
		    res.on('searchReference', function(referral) {
			    console.log('referral: ' + referral.uris.join());
				  });
				    res.on('error', function(err) {
					    console.error('error: ' + err.message);
						  });
						    res.on('end', function(result) {
							    console.log('status: ' + result.status);
								  });
								  });


console.log("After search");
client.unbind(function(err) {
  assert.ifError(err);
});
console.log("After unbind");
```

When I run this script, I see the console messages saying that the script started and after each of the connect, search, and unbind steps.
However, I don't get any message about the search results so what am I doing wrong?  are my search options not sufficient for the task of searching?

**Comment by UziTech** [comment]
add

```js
client.on("error", (err) => {
  assert.ifError(err);
});
client.on("setupError", (err) => {
  assert.ifError(err);
});
client.on("connectError", (err) => {
  assert.ifError(err);
});
client.on("connectTimeout", (err) => {
  assert.ifError(err);
});
client.on("connectRefused", (err) => {
  assert.ifError(err);
});
```

to see if there are any connection errors

**Comment by donraikes** [comment]
@UziTech  Thank you for the code.  I added it to my script and didn't get any errors displayed.

After all of my console.log statements were output, I did get some search results, but the long and short of it was that it didn't return any results.

I don't know what kind of ldap server my company is using, so I can't make my question any more specific.  

Based on the output, it didn't look like there were any errors encountered during the search just no results.

my code is as follows:

        const ldap = require('ldapjs');
        const assert = require('assert');
        
        // LDAP Connection Settings
        const server = "ldap.example.com";
        const userEmail = "don@example.com";
        const adSuffix = "dc=example,dc=com"; // test.com
        
        // Create client and bind to AD
        console.log("Prepared to create the ldapjs client");
        const client = ldap.createClient({
            url: `ldap://${server}`
        });
        client.on("error", (err) => {
          assert.ifError(err);
        });
        client.on("setupError", (err) => {
          assert.ifError(err);
        });
        client.on("connectError", (err) => {
          assert.ifError(err);
        });
        client.on("connectTimeout", (err) => {
          assert.ifError(err);
        });
        client.on("connectRefused", (err) => {
          assert.ifError(err);
        });
        console.log("Client created.");
        
        /*
        client.bind(userPrincipalName,password,err => {
            assert.ifError(err);
        });
        */
        // Search AD for user
        const searchOptions = {
            scope: "sub",
            filter: `(email=${userEmail})`
        };
        
        console.log("Performing search...");
        client.search(adSuffix,searchOptions,(err,res) => {
            assert.ifError(err);
        
            res.on('searchEntry', entry => {
        		console.log("Search results:");
                console.log(entry.object);
            });
            res.on('searchReference', referral => {
        		console.log("Referral:");
                console.log('referral: ' + referral.uris.join());
            });
            res.on('error', err => {
        		console.log("Error during search");
                console.error('error: ' + err.message);
            });
            res.on('end', result => {
                console.log(result);
            });
        });
        console.log("After search.");
        // Wrap up
        client.unbind( err => {
            assert.ifError(err);
        });
        console.log("After unbind");


Is there anything in particular I should look for in the search result object to help determine why I got no usable results?

an excerpt from the search results:
        SearchResponse {
          messageID: 1,
          protocolOp: 101,
          controls: [],
          log: { child: [Function] },
          status: 0,
          matchedDN: '',
          errorMessage: '',
          referrals: [],

**Comment by donraikes** [comment]
I discovered by looking at the php code I am replacing that I constructed the search filter incorrectly.  Once I fixed that, I did get a lot of information in the response object.

The thing I still find interesting is that the console.log message stating that I was unbinding was printed before the search results were printed.

Is there a way to wait until I get a response to proceed or do I just need to add code into the 

        res.on('searchEntry', entry => {

function to process the results once they arrive?
